### ISSUE TITLE
Hash collision in CSS rule hashing

### ISSUE THREAD

**Issue by marvinhagemeister** [styled-components/styled-components#3561]
## Environment

`npx envinfo --system --binaries --npmPackages styled-components,babel-plugin-styled-components --markdown --clipboard`

System:
 - OS: macOS 11.5
 - CPU: (12) x64 Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz
 - Memory: 2.19 GB / 32.00 GB
 - Shell: 5.8 - /bin/zsh

Binaries:
 - Node: 14.16.1 - ~/.nvm/versions/node/v14.16.1/bin/node
 - Yarn: 1.22.10 - /usr/local/bin/yarn
 - npm: 6.14.12 - ~/.nvm/versions/node/v14.16.1/bin/npm

npmPackages:
 - babel-plugin-styled-components: ^1.13.2 => 1.12.0 
 - styled-components: ^5.3.0 => 5.3.0 

## Steps to reproduce

Very hard to reproduce in isolation. We tried to narrow it down in a client project, but it was only visible in a full environment with nextjs and SSR at play.

## Expected Behavior

Elements created via styled components should receive the CSS rules they were created with.

## Actual Behavior

In our setup a styled component suddenly receives CSS from another component. In our case it happened when we hydrate an SSR'ed HTML in the client. Upon further inspection it turns out that the same hash is already present in the ruleset collection. This leads to the wrong className being added to the current element. The source of this problem is a fault in the hashing code.

The actual hashing function that is shipped in the npm package for `5.3.0` contains an error due to which most of the hashing code is bypassed. The source code looks like this:

```js
// simplified:
generateAndInjectStyles(...) {
  //...snip
  const cssStatic = "color: red" // example
  // Note that the second parameter is a string of the CSS
  const name = generateName(phash(this.baseHash, cssStatic) >>> 0);
}

const phash = (h, x) => {
  let i = x.length;

  while (i) {
    h = (h * 33) ^ x.charCodeAt(--i);
  }

  return h;
};
```

And this is the generated minified code from the above:

```js
var t = ke(this.rules, e, a, n).join("")
  // Second parameter should be a string, namely just `t` and not `t.length`
  , o = W(q(this.baseHash, t.length) >>> 0);

// q is phash()
var q = function(e, a) {
  // Parameter "a" will always be a number. Therefore the test expression
  // inside the for construct will ALWAYS be falsy. In effect the actual
  // hashing logic is skipped and we'll just return parameter "e" as is.
  for (var n = a.length; n; )
    e = 33 * e ^ a.charCodeAt(--n);
  return e
}
```

So in essence all the hashnig function is doing is this: 

```js
const phash = (h, x) => {
  return h;
};
```

The bug is here:

```diff
  var t = ke(this.rules, e, a, n).join("")
-   , o = W(q(this.baseHash, t.length) >>> 0);
+   , o = W(q(this.baseHash, t) >>> 0);
```

In summary the current hashing is bypassed and the hash is solely based on the concatenated CSS string length. This can easily lead to collisions.

The good news is that the issue is fixed in the current `main` branch. We weren't able to narrow it down to a particular commit as the git history seems to be composed of detached commits. Due to that `git bisect` doesn't work on this repo.

The fix can be verified by building the source in `main` and comparing the generated files in the `dist/` folder (just search for `>>>`, the hashing function is close by).

From my observation this seems to be the cause for these issues:

- https://github.com/styled-components/styled-components/issues/3482
- https://github.com/styled-components/styled-components/issues/3522

**Comment by kitten** [comment]
That's probably a good missing piece of debugging information for #3482. I'll make a branch with this fix applied and reverting the other change that's mentioned in #3482 and hopefully that'll take care of this

**Comment by marvinhagemeister** [comment]
@kitten Awesome, thank you so much for the quick fix. I'll check on Monday 👍

**Comment by marvinhagemeister** [comment]
Alright, can confirm that this resolves the issue we were experiencing. Each component has now the correct CSS properties 👍
